### PR TITLE
feat: skip status [OTE-349] [2/n]

### PR DESCRIPTION
**Pull request by yogurtandjam**
Followup to https://github.com/dydxprotocol/v4-abacus/pull/460

Fully unlocks the status tracking flow by calling the required `track` API call first before attempting to call `status`

- creates skipTrackProcessor
- adds new piece of abacus state trackStatuses
- simple bool switch to track whether or not a tx has been successfully tracked or not, so we can begin hitting it with `status`

### CODE REVIEW

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt [hunk 1]
```diff
@@ -1067,6 +1078,7 @@ open class TradingStateMachine(
         var configs = state?.configs
         var input = state?.input
         var transferStatuses = state?.transferStatuses?.toIMutableMap()
+        var trackStatuses = state?.trackStatuses?.toIMutableMap()
```
Comment (prashanDYDX) [inline src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt:1081]: why does this need to be an `IMutableMap`? (`IMutableMap` is only needed for types that are consumed directly by JavaScript, internally in abacus we can just use `Map` and `.toMap()`

File: src/commonTest/kotlin/exchange.dydx.abacus/tests/payloads/SkipTrackMock.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package exchange.dydx.abacus.tests.payloads
+
+class SkipTrackMock {
```
Comment (prashanDYDX) [inline src/commonTest/kotlin/exchange.dydx.abacus/tests/payloads/SkipTrackMock.kt:3]: this doesn't need to be a class, can just be an `object`. then you don't need to instantiate this in your test class. can just access like `SkipTrackMock`

File: src/commonMain/kotlin/exchange.dydx.abacus/output/PerpetualState.kt [hunk 3]
```diff
@@ -33,6 +33,7 @@ data class PerpetualState(
     val input: Input?,
     val availableSubaccountNumbers: IList<Int>,
     val transferStatuses: IMap<String, TransferStatus>?,
+    val trackStatuses: IMap<String, Boolean>?,
```
Comment (prashanDYDX) [inline src/commonMain/kotlin/exchange.dydx.abacus/output/PerpetualState.kt:36]: does this need to be nullable? does empty work here?

File: src/commonTest/kotlin/exchange.dydx.abacus/tests/payloads/SkipTrackMock.kt (hunk 2 above)
Comment (yogurtandjam) [reply src/commonTest/kotlin/exchange.dydx.abacus/tests/payloads/SkipTrackMock.kt:3]: got it - though FYI i've been following existing patterns so we should probably refactor existing mocks

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt (hunk 1 above)
Comment (yogurtandjam) [reply src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt:1081]: i was under the impression that this state is exposed to the js client, no? this is the `update` fn that constructs the `PerpetualState` object so all the maps are `IMap`s

File: src/commonMain/kotlin/exchange.dydx.abacus/output/PerpetualState.kt (hunk 3 above)
Comment (yogurtandjam) [reply src/commonMain/kotlin/exchange.dydx.abacus/output/PerpetualState.kt:36]: hmm... maybe. right now the abacus pattern is to return null for invalid data but we can probably change the default value to an empty map. i'll attempt to do it and timebox it for ~30 mins

Comment (yogurtandjam) [reply src/commonMain/kotlin/exchange.dydx.abacus/output/PerpetualState.kt:36]: ah - everything is pulled from the typeless `data` map so everything **has** to be nullable until we get state schematized

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt (hunk 1 above)
Comment (yogurtandjam) [reply src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt:1068]: or are we intentionally selective about what abacus state items get exposed to the JS client

File: src/commonMain/kotlin/exchange.dydx.abacus/output/PerpetualState.kt [hunk 4]
```diff
@@ -119,4 +120,8 @@ data class PerpetualState(
     fun transferStatus(hash: String): TransferStatus? {
         return transferStatuses?.get(hash)
     }
+
+    fun trackStatus(hash: String): Boolean? {
+        return trackStatuses?.get(hash)
+    }
```
Comment (mike-dydx) [inline src/commonMain/kotlin/exchange.dydx.abacus/output/PerpetualState.kt:126]: is this just a convenience method for map access? Ideally we access maps uniformly throughout the app IMO

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+Squid.kt [hunk 5]
```diff
@@ -87,3 +87,16 @@ internal fun TradingStateMachine.squidStatus(
...
+
+internal fun TradingStateMachine.routerTrack(
+    payload: String,
+): StateChanges? {
+    val json = try {
+        Json.parseToJsonElement(payload).jsonObject.toMap()
+    } catch (exception: SerializationException) {
+        Logger.e { "Failed to deserialize skipTrack: $payload \nException: $exception" }
+        return StateChanges(iEmptyList())
+    }
```
Comment (mike-dydx) [inline src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+Squid.kt:99]: maybe can use ParserProtocol's

`decodeJsonObject` to reduce this bit

File: src/commonMain/kotlin/exchange.dydx.abacus/state/v2/supervisor/OnboardingSupervisor.kt [hunk 6]
```diff
@@ -1158,6 +1163,48 @@ internal class OnboardingSupervisor(
...
+            return
+        }
+        val params: IMap<String, String> = iMapOf(
+            "tx_hash" to hash,
+            "chain_id" to fromChainId,
+        ).filterNotNull()
+        val url = helper.configs.skipV2Status()
+        helper.get(url, params) { _, response, httpCode, _ ->
+            if (response != null) {
+                update(stateMachine.squidStatus(response, hash), oldState)
```
Comment (mike-dydx) [inline src/commonMain/kotlin/exchange.dydx.abacus/state/v2/supervisor/OnboardingSupervisor.kt:1184]: squidstatus?

Comment (yogurtandjam) [reply src/commonMain/kotlin/exchange.dydx.abacus/state/v2/supervisor/OnboardingSupervisor.kt:1184]: man, i think this is a rebase issue. this is an old PR. thanks

File: src/commonTest/kotlin/exchange.dydx.abacus/tests/payloads/SkipTrackMock.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package exchange.dydx.abacus.tests.payloads
+
+object SkipTrackMock {
+    val error = """{
+  "code": 5,
+  "message": "chain id not supported",
+  "details": []
```
Comment (ruixhuang) [inline src/commonTest/kotlin/exchange.dydx.abacus/tests/payloads/SkipTrackMock.kt:7]: Do we want to surface the error to the user? @mike-dydx

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+Squid.kt (hunk 5 above)
Comment (yogurtandjam) [reply src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+Squid.kt:99]: decodeJsonObject performs an extra conversion to `IMap` which is slower than the native kotlin map. if you want i can make a `decodeJsonObjectToNativeMap` or something

File: src/commonTest/kotlin/exchange.dydx.abacus/tests/payloads/SkipTrackMock.kt (hunk 7 above)
Comment (yogurtandjam) [reply src/commonTest/kotlin/exchange.dydx.abacus/tests/payloads/SkipTrackMock.kt:7]: FWIW i'm just following the same error pattern we have for other router requests for now.

In the future for error messages we'll be making an abstraction layer to make these more end-user friendly

**Review by ruixhuang — APPROVED** [review]

**Pull request merged**